data_structure/Binary heap detail implements priority queues.md

```diff
@@ -0,0 +1,213 @@
+    public Key delMax() {...}
+
+    /* swim the KTH element to maintain the maximum heap properties */
+    private void swim(int k) {...}
+
+    /* Sink the KTH element to maintain maximum heap properties */
+    private void sink(int k) {...}
+
+    /* Swap the two elements of the array */
+    private void exch(int i, int j) {
+        Key temp = pq[i];
+        pq[i] = pq[j];
+        pq[j] = temp;
+    }
+
+    /* Is pq[i] less than pq[j]？ */
+    private boolean less(int i, int j) {
+        return pq[i].compareTo(pq[j]) < 0;
+    }
+
+    /* and left, right, parent methods */
+}
+```
+
+The four methods that are left open are the mystery of binary heap and priority queue, which are illustrated below.
+
+### 3、Implement swim and sink
+
+Why swim up and sink? To maintain the heap structure!
+
+We're talking about a maximum heap, where each node is larger than its two children, but when you insert and delete elements, you break the nature of the heap, which requires both operations to restore the nature of the heap.
+
+For a maximum heap, there are two cases where the nature of the heap is destroyed:
+
+1. If some node A is smaller than its children, then A doesn't deserve to be the parent node, so it should go down, and the larger node down here comes up as the parent node, and that's  **sink** on A。
+
+2. If some node A is bigger than its parent, then A should not be the child node, but the parent node should be replaced and the parent node should be the parent node itself, which is the **swim** of A。
+
+Of course, the misaligned node A may have to rise (or sink) many times before it reaches the correct location and restores the nature of the heap. So there must be a  `while`  loop in the code.
+
+The observant reader may ask, aren't these two operations reciprocal, so the operation of floating up must be accomplished by sinking down, why should I bother to write two methods?
+
+Yes, the operations are inversely equivalent, but in the end we will only operate at the bottom and the top of the heap (we'll see why), and obviously the "misaligned" elements at the bottom of the heap need to float up, and the "misaligned" elements at the top of the heap need to sink.
+
+**上浮的代码实现：**
+
+```java
+private void swim(int k) {
+    // If you float to the top of the heap, you can't float any higher
+    while (k > 1 && less(parent(k), k)) {
+        // If the KTH element is larger than the top
+        // replace k
+        exch(parent(k), k);
+        k = parent(k);
+    }
+}
+```
+
+Draw a GIF to help you understand：
+
+![2](../pictures/heap/swim.gif)
+
+**Sinking code implementation：**
+
+Sinking is slightly more complicated than floating up, because floating up A node A only requires A to compare the size with its parent node; However, to sink A node A, we need to compare the size of A with its **two child nodes** If A is not the largest, we need to adjust the position and exchange the larger child node with A.
```

Comment:
floating up A node A only requires A 是不是多写了 A 呀，而且这个比较大小是不是 compare the value 比 compare the size 要清楚一点呢，可以把文中关于二叉堆节点值的【大小】翻译成 value

Reply:
好的

Reply:
Sinking is slightly more complicated than rising,Because some node A is floating up, all you need to do is compare the value of A with its parent,But sinking some node A,you need to compare the value of A with its two child nodes, and if A is not the largest, you need to adjust the position, and then swap the larger child node with A.   我准备提交时，你应经合并了，这是准备改变的地方，这是我第一次当贡献者，感觉very good，希望这个仓库越来越好